dashboard/test/controllers/aichat_controller_test.rb

```diff
@@ -72,16 +72,45 @@ class AichatControllerTest < ActionController::TestCase
     assert_response :bad_request
   end
 
-  # Post request with a profane messages param returns a failure
   test 'returns failure when chat message contains profanity' do
     sign_in(@genai_pilot_student)
     ShareFiltering.stubs(:find_failure).returns(ShareFailure.new(ShareFiltering::FailureType::PROFANITY, 'damn'))
     post :chat_completion, params: @profanity_violation_params, as: :json
     assert_equal ShareFiltering::FailureType::PROFANITY, json_response["status"]
     assert_equal "damn", json_response["flagged_content"]
+
+    session = AichatSession.find(json_response['session_id'])
+    stored_message = JSON.parse(session.messages)[0]
+    assert_equal stored_message, {
+      role: 'user',
+      content: @profanity_violation_message,
+      status: 'profanity_violation'
+    }.stringify_keys
+  end
+
+  test 'filters previous profanity when sending previous messages to Sagemaker but still logs' do
+    ok_message = {status: 'ok', role: 'user', content: 'another message'}.stringify_keys
+    params = @valid_params.merge(
+      storedMessages: [
+        {status: 'profanity_violation', role: 'user', content: 'damn'},
+        ok_message
+      ]
+    )
+
+    # Note that second expected argument filters out the previous profane message
+    # in what we send to Sagemaker.
+    AichatSagemakerHelper.expects(:format_inputs_for_sagemaker_request).
```

Comment:
Do you mind explaining again why you needed `to_unsafe_h` to test this function? I didn't quite understand from [your comment](https://github.com/code-dot-org/code-dot-org/blob/99a91995d5b86aa754ae5c5e4bc946d401683ed0/dashboard/app/controllers/aichat_controller.rb#L39) in `aichat_controller.rb`.

Reply:
Yeah...so if you're passing in params[:key] directly, you're getting an instance of `ActionController::Params`:

https://api.rubyonrails.org/classes/ActionController/Parameters.html#method-i-5B-5D

And if you try to set an `expects` on that argument, you get different instances, and the test fails.

You can compare hashes for equality, so `to_unsafe_h` allows you do that.

There certainly might be a better way to do this, but I couldn't figure it out. I thought you might be able to "record" the arguments passed to a stubbed function (then call `to_h` or something on the recorded argument and do a comparison), but couldn't figure out how to do that?

Here's some info on why to use "strong params" generally, which made me think in this case using `to_unsafe_h` isn't actually unsafe (we do it in a place or two already in our code base):

https://www.writesoftwarewell.com/why-use-strong-parameters-in-rails/

Reply:
Gotcha - thanks so much for the explanation!